File: Data Structures & Algorithms/longest-consecutive-sequence/submission-0.java
class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for(int n : nums){
            set.add(n);
        }
        int max = 0;
        Set<Integer> checked = new HashSet<>();
        for(int n : set){
            if(checked.contains(n)) continue;
            int counter = 1;
            for(int i = 1; set.contains(n - i); i++){
                counter++;
                checked.add(n - 1);
            }
            for(int i = 1; set.contains(n + i); i++){
                counter++;
                checked.add(n + 1);
            }
            if(counter > max){
                max = counter;
            }
        }
        return max;
    }
}

File: Data Structures & Algorithms/course-schedule/submission-0.kt
class Solution {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        val head = GraphNode()
        val map = HashMap<Int, GraphNode>().apply{ put(0, head) }
        for(i in 1 until numCourses){
            map.put(i, GraphNode())
        }
        for(pre in prerequisites){
            map[pre[0]]?.let{
                it.neighbors.add(map[pre[1]]!!)
            }
        }
        fun hasLoop(node : GraphNode, checked : HashSet<GraphNode>): Boolean{
            if(checked.contains(node)) return true
            checked.add(node)
            for(n in node.neighbors){
                if(hasLoop(n, checked)) return true
            }
            checked.remove(node)
            return false
        }
        val checked = HashSet<GraphNode>()
        for(n in map.values){
            if(hasLoop(n, checked)) return false
        }
        return true
    }
    class GraphNode{
        val neighbors = mutableListOf<GraphNode>()
    }
}
